Content.Server/Radiation/Systems/RadiationSystem.GridCast.cs

```diff
@@ -0,0 +1,239 @@
+        var radRay = new RadiationRay(mapId, sourceUid,sourceWorld,
+            destUid,destWorld, incomingRads)
+        {
+            Grid = gridUid,
+            VisitedTiles = visitedTiles
+        };
+
+        // if grid doesn't have resistance map just apply distance penalty
+        if (!resistanceQuery.TryGetComponent(gridUid, out var resistance))
+            return radRay;
+        var resistanceMap = resistance.ResistancePerTile;
+
+        // get coordinate of source and destination in grid coordinates
+        // todo: entity queries doesn't support interface - use it when IMapGridComponent will be removed
+        if (!TryComp(gridUid, out IMapGridComponent? grid))
+            return radRay;
+        var sourceGrid = grid.Grid.TileIndicesFor(sourceWorld);
+        var destGrid = grid.Grid.TileIndicesFor(destWorld);
+
+        // iterate tiles in grid line from source to destination
+        var line = Line(sourceGrid.X, sourceGrid.Y, destGrid.X, destGrid.Y);
+        foreach (var point in line)
+        {
+            if (resistanceMap.TryGetValue(point, out var resData))
+            {
+                radRay.Rads -= resData;
+                visitedTiles.Add((point, radRay.Rads));
+            }
+            else
+            {
+                visitedTiles.Add((point, null));
+            }
+
+            // no intensity left after blocker
+            if (radRay.Rads <= MinIntensity)
+            {
+                radRay.Rads = 0;
+                return radRay;
+            }
+        }
+
+        return radRay;
+    }
+
+    private RadiationRay Raycast(MapId mapId, EntityUid sourceUid, EntityUid destUid,
+        Vector2 sourceWorld, Vector2 destWorld, Vector2 dir, float distance, float incomingRads,
+        EntityQuery<RadiationBlockerComponent> blockerQuery)
+    {
+        var blockers = new List<(Vector2, float)>();
+        var radRay = new RadiationRay(mapId, sourceUid, sourceWorld,
+            destUid, destWorld, incomingRads)
+        {
+            Blockers = blockers
+        };
+
+        var colRay = new CollisionRay(sourceWorld, dir, (int) CollisionGroup.Impassable);
+        var results = _physicsSystem.IntersectRay(mapId, colRay, distance, returnOnFirstHit: false);
+
+        foreach (var obstacle in results)
+        {
+            if (!blockerQuery.TryGetComponent(obstacle.HitEntity, out var blocker))
+                continue;
+
+            radRay.Rads -= blocker.RadResistance;
+            blockers.Add((obstacle.HitPos, radRay.Rads));
+
+            if (radRay.Rads <= MinIntensity)
+            {
+                radRay.Rads = 0;
+                return radRay;
+            }
+        }
+
+        return radRay;
+    }
+
+
+    // bresenhams line algorithm
+    // this is slightly rewritten version of code bellow
+    // https://stackoverflow.com/questions/11678693/all-cases-covered-bresenhams-line-algorithm
+    private IEnumerable<Vector2i> Line(int x, int y, int x2, int y2)
```

Comment:
Regarding sloth's other comment about not using IEnumerables, does changing to an enumerator or directly including this in the loop at the cost of readability appreciably help performance? Because even if the code is less readable it might be worth it.

Reply:
1. I tried to include code directly into loop and didn't see any consistent performance changes in benchmark scenario. Maybe  it's possible to push it to the limit, but I just doesn't see any real in game situation where this will become bottleneck.
2. Ideally this code should be available for other systems that want to do line grid traversal. So IEnumerable seems to be reasonable option to make this code reusable.

Reply:
Did you run the benchmark on debug or release + did you check allocations? A struct enumerator will pretty much always be better especially in hot loops, it just requires more effort. Most engine grid methods internally use them now, e.g. for iterating tiles.

Reply:
So I cranked system to the limit. This is about 1k rays through station. After that in release mode there difference between IEnumerator and embedded into loop showed up.

IEnumerator:
```
[INFO] root: Radiation update: 1.9937ms with. Receivers: 58, Sources: 18, Rays: 985
[INFO] root: Radiation update: 2.0724ms with. Receivers: 58, Sources: 18, Rays: 985
[INFO] root: Radiation update: 1.9457ms with. Receivers: 58, Sources: 18, Rays: 985
[INFO] root: Radiation update: 2.0448ms with. Receivers: 58, Sources: 18, Rays: 985
[INFO] root: Radiation update: 1.9511ms with. Receivers: 58, Sources: 18, Rays: 985
[INFO] root: Radiation update: 1.971ms with. Receivers: 58, Sources: 18, Rays: 985
[INFO] root: Radiation update: 1.9301ms with. Receivers: 58, Sources: 18, Rays: 985
```

Inside update loop:
```
[INFO] root: Radiation update: 1.4641ms with. Receivers: 58, Sources: 18, Rays: 985
[INFO] root: Radiation update: 1.5862ms with. Receivers: 58, Sources: 18, Rays: 985
[INFO] root: Radiation update: 1.5702ms with. Receivers: 58, Sources: 18, Rays: 985
[INFO] root: Radiation update: 1.5564ms with. Receivers: 58, Sources: 18, Rays: 985
[INFO] root: Radiation update: 1.6393ms with. Receivers: 58, Sources: 18, Rays: 985
[INFO] root: Radiation update: 1.5513ms with. Receivers: 58, Sources: 18, Rays: 985
[INFO] root: Radiation update: 1.5575ms with. Receivers: 58, Sources: 18, Rays: 985
```

Let me check how struct enumerator behaves.

Reply:
Well, thats impressive. Struct enumerators indeed much faster than IEnumerator. From my tests it's about same speed as code placed inside loop. I prepare PR for engine here:
https://github.com/space-wizards/RobustToolbox/pull/3278